### PR TITLE
Optimisations

### PR DESCRIPTION
**Pull request by colinleroy**
Hello,

I hope you'll consider this PR. It aims at shaving a bit of performance (about 20% in redis-cache on my homepage) with two modifications:
- Avoid walking full arrays to determine group type: We use an associative array to get the type of the group instead of walking an array.
- Avoid sanitizing key/group multiple times in the same codepath: for example, ->get() was using sanitize_key to build the key and to check if the group was global.

I'll attach screenshots of KCachegrind that demonstrate the changes.

Hope this helps,
Colin

### CODE REVIEW

**Comment by colinleroy** [comment]
Here is the call graph on WP_Object_Cache->get() before:
image

Here it is after:
image

**Comment by colinleroy** [comment]
Here is the total time passed in WP_Object_Cache functions before:
image

Here it is after:
image

You can see an improvement of about 19%.

File: includes/object-cache.php [hunk 1]
```diff
@@ -1175,10 +1198,14 @@ protected function add_or_replace( $add, $key, $value, $group = 'default', $expi
...
-        $derived_key = $this->build_key( $key, $group );
+
+        $san_key = $this->sanitize_key_part( $key );
+        $san_group = $this->sanitize_key_part( $group );
+
+        $derived_key = $this->fast_build_key( $san_key, $san_group );
 
         // Save if group not excluded and redis is up.
-        if ( ! $this->is_ignored_group( $group ) && $this->redis_status() ) {
+        if ( ! $this->is_group_of_type( $san_group, "ignored" ) && $this->redis_status() ) {
```
Comment (tillkruss) [inline includes/object-cache.php:1208]: For readability, let's keep `is_ignored_group()` and  `is_global_group()`.

**Review by tillkruss — CHANGES_REQUESTED** [review]

**Comment by colinleroy** [comment]
Hi @tillkruss ,
Thank you for your feedback. I've addressed your comment.
I also added a part in diagnostics to check on that array, and added refreshing the group type cache in case someone calls add_..._groups().

**Comment by colinleroy** [comment]
Hi @tillkruss! Do you need me to fix something else on this PR?
Thanks :)
Colin

**Comment by tillkruss** [comment]
@colinleroy: I have to review the changes in detail to make sure nothing breaks, otherwise 100,000 sites break.

**Comment by colinleroy** [comment]
Ah, yes, understandably :)
Thanks !

**Comment by tillkruss** [comment]
@colinleroy: Can you drop me an email via the chat bubble on https://objectcache.pro?

**Comment by colinleroy** [comment]
Hi @tillkruss,

I'm sorry I've not been able to really get in touch with you yet... Life's a bit hectic. 
Do you still want to apply this PR ?

Thanks!
Colin

**Comment by tillkruss** [comment]
@colinleroy: Yeah, this looks good.

**Pull request merged**